mlflow/projects/__init__.py

```diff
@@ -1,156 +1,88 @@
-        dest_path = os.path.join(storage_dir, basename)
-        if dest_path != user_param_value:
-            data.download_uri(uri=user_param_value, output_path=dest_path)
-        return os.path.abspath(dest_path)
-
-    def compute_value(self, user_param_value, storage_dir):
-        if self.type != "path" and self.type != "uri":
-            return user_param_value
-        if self.type == "uri":
-            return self._compute_uri_value(user_param_value)
-        return self._compute_path_value(user_param_value, storage_dir)
-
-
-def _sanitize_param_dict(param_dict):
-    return {str(key): shlex_quote(str(value)) for key, value in param_dict.items()}
+def run(uri, entry_point="main", version=None, parameters=None, experiment_id=None,
+        mode=None, cluster_spec=None, git_username=None, git_password=None, use_conda=True,
+        use_temp_cwd=False, storage_dir=None, block=True):
+    """
+    Run an MLflow project from the given URI.
+
+    Supports downloading projects from Git URIs with a specified version, or copying them from
+    the file system. For Git-based projects, a commit can be specified as the `version`.
+
+    :param entry_point: Entry point to run within the project. If no entry point with the specified
+                        name is found, attempts to run the project file `entry_point` as a script,
+                        using "python" to run .py files and the default shell (specified by
+                        environment variable $SHELL) to run .sh files.
+    :param experiment_id: ID of experiment under which to launch the run.
+    :param mode: Execution mode for the run. Can be set to "local" or "databricks".
+    :param cluster_spec: Path to JSON file describing the cluster to use when launching a run on
+                         Databricks.
+    :param git_username: Username for HTTP(S) authentication with Git.
+    :param git_password: Password for HTTP(S) authentication with Git.
+    :param use_conda: If True (the default), creates a new Conda environment for the run and
+                      installs project dependencies within that environment. Otherwise, runs the
+                      project in the current environment without installing any project
+                      dependencies.
+    :param use_temp_cwd: Only used if `mode` is "local" and `uri` is a local directory.
+                         If True, copies project to a temporary working directory before running it.
+                         Otherwise (the default), runs project using `uri` (the project's path) as
+                         the working directory.
+    :param storage_dir: Only used if `mode` is local. MLflow will download artifacts from
+                        distributed URIs passed to parameters of type 'path' to subdirectories of
+                        storage_dir.
+    :param block: Whether or not to block while waiting for a run to complete. Defaults to True.
+    :return: UUID of the run
```

Comment:
We return run UUIDs from `projects.run`, but it could be better to instead return an instance of a wrapper class that knows the store corresponding to the run and exposes a `get_run` method that fetches run info from that store.

This would be useful e.g. when launching runs on Databricks (especially if we automatically set the tracking URI to point to the Databricks tracking server); currently users don't have a great way of getting run info from a non-local store / have to know which store corresponds to a run.

Reply:
I don't think we anticipate clients connecting to multiple stores; in production, it should always be a Databricks/remote tracking server, right? So not having the information returned by this method seems fine.

Reply:
Yeah agreed, can only think of pretty esoteric use cases where I'd want to launch runs against different tracking servers (i.e. if I'm writing a cron script that launches some runs on Databricks, some runs in another environment / against a different tracking server, etc)

Reply:
Seems fine to return a run_uuid for now

Reply:
I've thought about this and I believe it's better to return a SubmittedRun wrapper class that has a run_id field. (BTW let's call it run_id throughout the public APIs in the future, not run_uuid.) Later on, we can add other features to this class without impacting user code, such as a method to wait for the run to finish. Otherwise, if we return just the ID now, there's no easy way to add these.

There's also an argument that people could do this by calling the Tracking API directly on the ID we returned though, but to me it seems easier to give them an object that they can call wait or whatever on. What do you guys think? Maybe @mparkhe and @andrewmchen have thoughts on this too.

Reply:
+1 to returning context object, that can be used with `with` semantics, with some limited support akin to `tracking.start_run()` which returns `ActiveRun` -- if that can support `log_params`, `log_metrics`, ... etc that proxies it over to `mlflow.log...` API.

Reply:
Yeah adding a wrapper class seems good to me, to give us flexibility maybe we can just return a wrapper around run ID for now & add more APIs in the future - I do see use cases for the things @mparkhe mentioned, e.g. launching a Databricks run and then adding tags/parameters to it containing a link to the run result page, etc.